### PR TITLE
Tweak: One click site - Import base [ED-5813]

### PR DESCRIPTION
**Pull request by guyc-E**

### CODE REVIEW

File: core/app/modules/import-export/directories/base.php [hunk 1]
```diff
@@ -95,9 +120,22 @@ final public function run_import( array $settings ) {
 			$meta_data[ $sub_directory_name ] = $sub_directory->run_import( $settings[ $sub_directory_name ] );
 		}
 
+		$is_root = array_key_exists( 'site-settings', $settings );
```
Comment (matipojo) [inline core/app/modules/import-export/directories/base.php:123]: There is a Root class? If no, try to find a better way to indicate the root.

File: core/app/modules/import-export/directories/base.php [hunk 2]
```diff
@@ -50,6 +55,22 @@ public function __construct( Iterator $iterator, Base $parent = null ) {
...
+			$this->set_documents_elements_by_id( $document->get_main_id(), $data['elements'] );
+
+			$data['elements'] = [];
+		}
+
+		return $data;
+	}
+
+	private function set_documents_elements_by_id( $id, $elements ) {
+		self::$documents_elements[ $id ] = $elements;
```
Comment (matipojo) [inline core/app/modules/import-export/directories/base.php:73]: `self` is shared with all extends of the base.
Can you replace across the file?
```suggestion
		static::$documents_elements[ $id ] = $elements;
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/app/modules/import-export/directories/base.php (hunk 1 above)
Comment (guyc-E) [reply core/app/modules/import-export/directories/base.php:123]: Fixed

File: core/app/modules/import-export/directories/base.php (hunk 2 above)
Comment (guyc-E) [reply core/app/modules/import-export/directories/base.php:73]: I need it to be shared with all extended classes.
The $this keyword in line 125 is different in each iteration.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/app/modules/import-export/import.php [hunk 3]
```diff
@@ -26,7 +28,23 @@ final public function run() {
...
+		$results = $root_directory->run_import( $manifest_data );
+
+		$this->save_elements_of_imported_posts( $results );
+
+		return $results;
+	}
+
+	public function prevent_saving_elements_on_post_creation( $data, $document ) {
+		if ( $data['elements'] ) {
+			$this->$documents_elements[ $document->get_main_id() ] = $data['elements'];
```
Comment (matipojo) [inline core/app/modules/import-export/import.php:42]:
```suggestion
			$this->documents_elements[ $document->get_main_id() ] = $data['elements'];
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/app/modules/import-export/import.php [hunk 4]
```diff
@@ -59,4 +77,23 @@ private function init_adapters( array $manifest_data ) {
...
+			$map_old_new_post_ids = $results['succeed'];
+		}
+
+		foreach( $results as $result ) {
+			if ( isset( $result['succeed'] ) ) {
+				$map_old_new_post_ids += $result['succeed'];
+			}
+		}
+
+		foreach ( $this->$documents_elements as $new_id => $document_elements ) {
```
Comment (matipojo) [inline core/app/modules/import-export/import.php:94]:
```suggestion
		foreach ( $this->documents_elements as $new_id => $document_elements ) {
```

File: core/app/modules/import-export/import.php [hunk 5]
```diff
@@ -59,4 +78,29 @@ private function init_adapters( array $manifest_data ) {
...
 	}
+
+	private function save_elements_of_imported_posts( $results ) {
+		$map_old_new_post_ids = [];
+
+		$map_old_new_post_ids = $this->map_old_new_post_ids( $results );
+
+		foreach ( $this->documents_elements as $new_id => $document_elements ) {
+			$document = Plugin::$instance->documents->get( $new_id );
+			$document->on_import_replace_dynamics_elements_id( $document_elements, $map_old_new_post_ids );
```
Comment (matipojo) [inline core/app/modules/import-export/import.php:89]:
```suggestion
			$document->on_import_replace_dynamic_content( $document_elements, $map_old_new_post_ids );
```

File: core/dynamic-tags/base-tag.php [hunk 6]
```diff
@@ -80,12 +80,22 @@ public function is_settings_required() {
 		return false;
 	}
 
+	/**
+	 * Each tag should replace its own dynamic data.
+	 *
+	 * @since 3.6.0
+	 * @access public
+	 */
+	public function on_import_replace_dynamic_tag( $map_old_new_post_ids ) {
```
Comment (matipojo) [inline core/dynamic-tags/base-tag.php:89]:
```suggestion
	public function on_import_replace_dynamic_content( $map_old_new_post_ids ) {
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/base/document.php [hunk 7]
```diff
@@ -1065,6 +1066,28 @@ public function delete() {
...
 
+	/**
+	 *
+	 * @since 3.6.0
+	 *
+	 * @param array $elements
+	 *
+	 * @param array $map_old_new_post_ids
+	 */
+	public function on_import_replace_dynamic_content( &$elements, $map_old_new_post_ids ) {
```
Comment (matipojo) [inline core/base/document.php:1077]:
```suggestion
	public function on_import_replace_dynamic_content( $elements, $map_old_new_post_ids ) {
```

File: core/base/document.php [hunk 8]
```diff
@@ -1065,6 +1066,28 @@ public function delete() {
...
+			$element_instance = Plugin::$instance->elements_manager->create_element_instance( $element_config );
+
+			if ( $element_instance ) {
+				$element_config = $element_instance->on_import_replace_dynamic_content( $element_config, $map_old_new_post_ids );
+
+				$this->on_import_replace_dynamic_content( $element_config['elements'], $map_old_new_post_ids );
+			}
+		}
+
+		$this->save_elements( $elements );
```
Comment (matipojo) [inline core/base/document.php:1088]: Return the updated config, like the base method.
Save the elements from the caller `save_elements_of_imported_posts`

File: includes/base/element-base.php [hunk 9]
```diff
@@ -573,6 +573,30 @@ public function is_type_instance() {
...
+	 *
+	 * @since 3.6.0
+	 * @access public
+	 *
+	 * @return array Element data.
+	 */
+	public function on_import_replace_dynamic_content( $config, $map_old_new_post_ids ) {
+		$tags_manager = Plugin::$instance->dynamic_tags;
+
+		foreach ( $config['settings'][ $tags_manager::DYNAMIC_SETTING_KEY ] as $dynamic_name => $dynamic_value ) {
```
Comment (matipojo) [inline includes/base/element-base.php:587]: Check if isset `$config['settings'][ $tags_manager::DYNAMIC_SETTING_KEY ]`

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/base/document.php [hunk 10]
```diff
@@ -1065,6 +1066,28 @@ public function delete() {
...
 
+	/**
+	 *
+	 * @since 3.6.0
+	 *
+	 * @param array $elements
+	 *
+	 * @param array $map_old_new_post_ids
+	 */
+	public static function on_import_replace_dynamic_content( $elements, $map_old_new_post_ids ) {
```
Comment (matipojo) [inline core/base/document.php:1077]:
```suggestion
	public static function on_import_replace_dynamic_content( $config, $map_old_new_post_ids ) {
```

File: core/base/document.php [hunk 11]
```diff
@@ -1065,6 +1066,28 @@ public function delete() {
...
+		foreach ( $elements as &$element_config ) {
+			$element_instance = Plugin::$instance->elements_manager->create_element_instance( $element_config );
+
+			if ( $element_instance ) {
+				$element_config = $element_instance->on_import_replace_dynamic_content( $element_config, $map_old_new_post_ids );
```
Comment (matipojo) [inline core/base/document.php:1082]:
```suggestion
				$element_config = $element_instance->on_import_replace_dynamic_content( $element_config, $map_old_new_post_ids );
```
```suggestion
				$element_config = $element_instance::on_import_replace_dynamic_content( $element_config, $map_old_new_post_ids );
```

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — APPROVED** [review]
🔥 Nice Work!

**Review by matipojo — CHANGES_REQUESTED** [review]
Please fix the checks errors

**Review by matipojo — APPROVED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
Please resolve the conflicts

**Pull request closed without merging**